federatedscope/attack/auxiliary/poisoning_data.py

```diff
@@ -0,0 +1,302 @@
+import torch
+from PIL import Image
+import numpy as np
+from torchvision.datasets import MNIST, EMNIST, CIFAR10
+from torchvision.datasets import DatasetFolder
+from torchvision import transforms
+from federatedscope.core.auxiliaries.transform_builder import get_transform
+from federatedscope.attack.auxiliary.backdoor_utils import selectTrigger
+from torch.utils.data import DataLoader, Dataset
+from federatedscope.attack.auxiliary.backdoor_utils import normalize
+import matplotlib
+import pickle
+import logging
+import os
+
+logger = logging.getLogger(__name__)
+
+
+def load_poisoned_dataset_edgeset(data, ctx, mode):
+
+    transforms_funcs = get_transform(ctx, 'torchvision')['transform']
+    load_path = ctx.attack.edge_path
+    if "femnist" in ctx.data.type:
+        if mode == 'train':
+            fraction = 0.1
+            train_path = os.path.join(
+                load_path, "poisoned_edgeset_fraction_{}".format(fraction))
+            with open(train_path, "rb") as saved_data_file:
+                poisoned_edgeset = torch.load(saved_data_file)
+            num_dps_poisoned_dataset = len(poisoned_edgeset)
+
+            for ii in range(num_dps_poisoned_dataset):
+                sample, label = poisoned_edgeset[ii]
+                # (channel, height, width) = sample.shape #(c,h,w)
+                sample = sample.numpy().transpose(1, 2, 0)
+                data['train'].dataset.append((transforms_funcs(sample), label))
+
+        if mode == 'test' or 'val':
+            poison_testset = list()
+            test_path = os.path.join(load_path, 'ardis_test_dataset.pt')
+            with open(test_path) as saved_data_file:
+                poisoned_edgeset = torch.load(saved_data_file)
+            num_dps_poisoned_dataset = len(poisoned_edgeset)
+
+            for ii in range(num_dps_poisoned_dataset):
+                sample, label = poisoned_edgeset[ii]
+                # (channel, height, width) = sample.shape #(c,h,w)
+                sample = sample.numpy().transpose(1, 2, 0)
+                poison_testset.append((transforms_funcs(sample), label))
+            data['poison_' + mode] = DataLoader(
+                poison_testset,
+                batch_size=ctx.data.batch_size,
+                shuffle=False,
+                num_workers=ctx.data.num_workers)
+
+    elif "CIFAR10" in ctx.data.type:
+        target_label = int(ctx.attack.target_label_ind)
+        target_label = 9
+        label = target_label
+        num_poisoned = 300
```

Comment:
target label & num_poisoned are not user defined?

Reply:
fixed it